pyro/infer/traceenum_elbo.py

```diff
@@ -261,6 +295,21 @@ def loss_and_grads(self, model, guide, *args, **kwargs):
         warn_if_nan(loss, "loss")
         return loss
 
+    def compute_marginals(self, model, guide, *args, **kwargs):
+        """
+        Computes marginal distributions at each model-enumerated sample site.
+
+        :returns: a dict mapping site name to marginal ``Distribution`` object
+        :rtype: OrderedDict
+        """
+        result = None
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
```

Comment:
This will still do guide enumeration (as opposed to sampling from previously guide-enumerated variables)

Reply:
Right. If we constrain this interface to be forwards-compatible with a future implementation that can enumerate guide sites, then this method will need to decide behavior based on `guide_trac.nodes[name]['infer']['enumerate']`. I believe the only forwards-compatible behavior is to raise a `NotImplementedError` if any guide site is set to be enumerated (I still need to write some of these error-checks in this PR).

If a user really wants guide-side enumeration during inference and guide sampling during prediction, then she can construct two different `TraceEnum_ELBO` classes; this will still work after we eventually implement querying for guide-enumeration-marginals.